File: src/serial_link.py
"""
Test forward dynamics using factor graphs.
Author: Frank Dellaert and Mandy Xie

We follow Lynch & Park 2017 conventions, but using j to index joints, as in Corke 2017:
    - j=0 is base
    - j in 1...N index joints
    - Mj is COM pose of link j in base frame 0.
    - Tj is the link frame j, aligned with joint j+1, expressed in base frame 0.
    - we use shorthand i==j-1 and k==j+1
"""

# pylint: disable=C0103, E1101, E0401, C0412

from __future__ import print_function

import gtsam
import numpy as np
import utils
from gtsam import Pose3, Rot3
from link import Link, a, V

ZERO6 = utils.vector(0, 0, 0, 0, 0, 0)


class SerialLink(object):
    """
        Calculate forward dynamics for manipulator using factor graph method
    """

    def __init__(self, calibration, base=Pose3(), tool=Pose3()):
        """ Constructor serial link manipulator from list of Link instances.
            Keyword arguments:
                calibration -- Link list
                base        -- optional wT0 base frame in world frame
                tool        -- optional tool frame in link N frame
        """
        self._links = calibration
        self._base = base
        self._tool = tool

        # Calculate screw axes for all joints, expressed in their COM frame.
        self._screw_axes = [link.screw_axis for link in self._links]

    @property
    def base(self):
        """Return base pose in world frame."""
        return self._base

    @property
    def tool(self):
        """Return tool pose in link N frame."""
        return self._tool

    @property
    def num_links(self):
        """Return number of *moving* links."""
        return len(self._links)

    def link_transforms(self, q=None):
        """ Calculate link transforms for all links.
            Keyword arguments:
                q (numpy array) -- optional joint angles (default all zero).
        """
        return [link.A(0 if q is None else q[i]) for i, link in enumerate(self._links)]

    def fkine(self, q):
        """ Forward kinematics.
            Keyword arguments:
                q (numpy array) -- joint angles.
            Returns tool frame in world frame.
        """
        t = self._base
        for A in self.link_transforms(q):
            t = t.compose(A)
        return t.compose(self._tool)

    def link_frames(self, q=None):
        """ Return each link frame for given joint angles.
            Note that frame Tj is aligned with the joint axis of joint j+1
            according to the Denavit-Hartenberg convention.
            Keyword arguments:
                q (numpy array) -- optional joint angles (default all zero).
            Returns tool frame in world frame.
        """
        frames = []
        t = self._base
        for A in self.link_transforms(q):
            t = t.compose(A)
            frames.append(t)
        return frames

    def com_frames(self, q=None):
        """ Return each link's center of mass frame at rest, in the world frame."""
        t = self._base
        frames = []
        for i, A in enumerate(self.link_transforms(q)):
            t = t.compose(A)
            iTcom = Pose3(Rot3(), self._links[i].center_of_mass)
            frames.append(utils.compose(t, iTcom))
        return frames

    def screw_axes(self):
        """Return screw axes for all joints, expressed in their COM frame."""
        return self._screw_axes

    def twists(self, Ts, joint_velocities):
        """ Return velocity twists for all joints, expressed in their COM frame.
            Keyword arguments:
                Ts -- com frames
                joint velocities (np.array, in rad/s)
        """
        # TODO(Frank): take jTi list instead, or better: do jTi with exp
        # The first link's twist is just from the joint
        twists = [self._screw_axes[0] * joint_velocities[0]]

        # Loop over joints j>1
        for j in range(2, self.num_links+1):
            # Equation 8.45 in MR, page 292
            twist_i = twists[-1]
            jTi = Ts[j-1].between(Ts[j-2])
            Aj = self._screw_axes[j-1]
            joint_vel_j = joint_velocities[j-1]
            twist_j = np.dot(jTi.AdjointMap(), twist_i) + Aj * joint_vel_j
            twists.append(twist_j)

        return twists

    def twists_gtsam(self, q, joint_velocities):
        """ Return velocity twists for all joints, expressed in their COM frame.
            This version uses GTSAM, mainly used to test twist factors.
            Keyword arguments:
                q (np.array, in rad) - joint angles
                joint velocities (np.array, in rad/s)
        """
        # configuration of link frame j-1 relative to link frame j for arbitrary joint angle
        jTis = self.jTi_list(q)

        gfg = gtsam.GaussianFactorGraph()

        # Add factor for each joint
        for i, (link, jTi, joint_vel_j) in enumerate(zip(self._links, jTis, joint_velocities)):
            gfg.add(link.twist_factor(i+1, jTi, joint_vel_j))

        result = gfg.optimize()
        return [result.at(V(j)) for j in range(1, self.num_links+1)]

    def jTi_list(self, q):
        """ Calculate list of transforms from COM frame j-1 relative to COM j.
            Keyword arguments:
                q (np.array, in rad) - joint angles
            Returns list of transforms, 2 more than number of links:
                - first transform is bT1, i.e. base expressed in link 1
                - last transform is tTnc, i.e., link N COM frame expressed in tool frame
        """
        # TODO(Frank): I don't like these inverses, wished we could do with forward.
        Ts = self.com_frames(q)
        bT1 = Ts[0].between(self._base)
        nTt = self.tool
        nTc = Pose3(Rot3(), self._links[-1].center_of_mass)
        tTnc = nTt.between(nTc)
        return [bT1] + [Tj.between(Ti) for Ti, Tj in zip(Ts[:-1], Ts[1:])] + [tTnc]

    def forward_factor_graph(self, q, joint_velocities, torques,
                             gravity_vector=None, base_twist_accel=ZERO6, external_wrench=ZERO6):
        """ Build factor graph for RR manipulator forward dynamics.
            Keyword arguments:
                q (np.array, in rad) - joint angles
                joint velocities (np.array, in rad/s)
                torques (np.array, in Nm)
                gravity_vector (np.array) -- if given, will create gravity forces
                base_twist_accel (np.array) -- optional acceleration to force at base
                external_wrench (np.array) -- optional external wrench
            Returns Gaussian factor graph
        """
        # TODO(Frank): take triples instead?
        N = self.num_links
        assert q.shape == (N,)
        assert joint_velocities.shape == (N,)
        assert torques.shape == (N,)

        # configuration of COM link frames
        Ts = self.com_frames(q)

        # Calculate all twists
        twists = self.twists(Ts, joint_velocities)

        # Set up Gaussian Factor Graph
        gfg = gtsam.GaussianFactorGraph()

        # Add factor to enforce base acceleration equal to zero
        gfg.add(Link.base_factor(base_twist_accel))

        # configuration of link frame j-1 relative to link frame j for arbitrary joint angle
        jTis = self.jTi_list(q)

        for i, (link, jTi, v_j, twist_j, torque_j, kTj) \
                in enumerate(zip(self._links, jTis, joint_velocities, twists, torques, jTis[1:])):
            j = i + 1
            factors = link.forward_factors(j, jTi, v_j, twist_j, torque_j, kTj, gravity_vector)
            gfg.push_back(factors)

        # Add factor to enforce external wrench at tool
        gfg.add(Link.tool_factor(self.num_links, external_wrench))

        return gfg

    def extract_joint_accelerations(self, result):
        """Extract joint accelerations for all joints from VectorValues."""
        return [result.at(a(j)) for j in range(1, self.num_links+1)]

    def factor_graph_optimization(self, forward_factor_graph):
        """ Optimize factor graph for manipulator forward dynamics.
            Keyword arguments: 
                forward_factor_graph -- factor graph for forward dynamics.
            Returns VectorValues with all unknowns:
                - N+1 twist accelerations (base + links)
                - N+1 torques (links + tool)
                - N joint accelerations.
            Note: use extract_joint_accelerations to filter out jojt accelerations.
        """
        return forward_factor_graph.optimize()

    def forward_dynamics(self, *args, **kwargs):
        """ Calculate joint accelerations from manipulator state and torques.
            See forward_factor_graph for input arguments.
        """
        factor_graph = self.forward_factor_graph(*args, **kwargs)
        result = factor_graph.optimize()
        return self.extract_joint_accelerations(result)

File: cpp/SerialLink.cpp
/**
 * @file  serial_link.cpp
 * @brief manipulator links
 * @Author: Frank Dellaert and Mandy Xie
 */

#include <DHLink.h>
#include <SerialLink.h>
#include <URDFLink.h>

using namespace std;
using namespace gtsam;

namespace manipulator {

template <typename T>
SerialLink<T>::SerialLink(const std::vector<T> &links, const Pose3 &base,
                          const Pose3 &tool)
    : links_(links), base_(base), tool_(tool) {
  // Calculate screw axes for all joints, expressed in their COM frame.
  for (auto &link : links_) {
    screwAxes_.push_back(link.screwAxis());
  }
}

template <typename T>
Vector SerialLink<T>::jointLowerLimits() const {
  Vector lower_limits = Vector::Zero(numLinks());
  for (int i = 0; i < numLinks(); ++i) {
    lower_limits[i] = links_[i].jointLowerLimit();
  }
  return lower_limits;
}

template <typename T>
Vector SerialLink<T>::jointUpperLimits() const {
  Vector upper_limits = Vector::Zero(numLinks());
  for (int i = 0; i < numLinks(); ++i) {
    upper_limits[i] = links_[i].jointUpperLimit();
  }
  return upper_limits;
}

template <typename T>
Vector SerialLink<T>::jointLimitThresholds() const {
  Vector limit_thresholds = Vector::Zero(numLinks());
  for (int i = 0; i < numLinks(); ++i) {
    limit_thresholds[i] = links_[i].jointLimitThreshold();
  }
  return limit_thresholds;
}

template <typename T>
vector<Pose3> SerialLink<T>::linkTransforms(const Vector &default_q) const {
  Vector q = default_q;
  if (equal(default_q, Vector::Zero(1))) {
    q = Vector::Zero(numLinks());
  }
  vector<Pose3> transforms;
  for (int i = 0; i < numLinks(); ++i) {
    transforms.push_back(links_[i].A(q[i]));
  }
  return transforms;
}

template <typename T>
vector<Pose3> SerialLink<T>::forwardKinematics(
    const Vector &q, boost::optional<std::vector<gtsam::Matrix> &> J) const {
  vector<Pose3> sTb;
  Pose3 t = base_;
  vector<Pose3> transforms = linkTransforms(q);
  for (auto const &A : transforms) {
    t = t * A;
    sTb.push_back(t);
  }
  sTb.push_back(t * tool_);

  if (J) {
    // body_manipulator_jacobian in body frame
    *J = bodyManipulatorJacobian(q, sTb);
  }

  return sTb;
}

template <typename T>
vector<Pose3> SerialLink<T>::linkFrames(const Vector &q) const {
  vector<Pose3> frames;
  Pose3 t = base_;
  vector<Pose3> transforms = linkTransforms(q);
  for (auto const &A : transforms) {
    t = t * A;
    frames.push_back(t);
  }
  return frames;
}

template <typename T>
vector<Pose3> SerialLink<T>::comFrames(const Vector &q) const {
  Pose3 t = base_;
  Pose3 iTcom;
  vector<Pose3> frames;
  vector<Pose3> transforms = linkTransforms(q);
  for (int i = 0; i < numLinks(); ++i) {
    t = t * transforms[i];
    iTcom = links_[i].centerOfMass();
    frames.push_back(t * iTcom);
  }
  return frames;
}

template <typename T>
vector<Pose3> SerialLink<T>::transformPOE(const Vector &default_q) const {
  Vector q = default_q;
  if (equal(default_q, Vector::Zero(1))) {
    q = Vector::Zero(numLinks());
  }
  vector<Pose3> frames;
  Pose3 frame = Pose3();
  frames.push_back(frame);
  vector<Vector6> screw_axes = spatialScrewAxes();
  for (int i = 0; i < numLinks(); ++i) {
    frame = frames.back() * Pose3::Expmap(screw_axes[i] * q[i]);
    frames.push_back(frame);
  }
  vector<Pose3> part(frames.begin() + 1, frames.end());
  return part;
}

template <typename T>
vector<Vector6> SerialLink<T>::spatialScrewAxes() const {
  vector<Vector6> spatial_screwAxes_vec;
  vector<Pose3> frames = comFrames();
  for (int i = 0; i < numLinks(); ++i) {
    spatial_screwAxes_vec.push_back(frames[i].AdjointMap() * screwAxes_[i]);
  }
  return spatial_screwAxes_vec;
}

template <typename T>
vector<Matrix> SerialLink<T>::spatialManipulatorJacobian(
    const Vector &q) const {
  // Calculate spatial_screw_axes at q
  vector<Vector6> screw_axes = spatialScrewAxes();
  // Calculate joint frames expressed in world frame
  vector<Pose3> transforms = transformPOE(q);
  vector<Matrix> Js(numLinks() + 1, Matrix::Zero(6, numLinks()));

  Matrix Jst_eef = Matrix::Zero(6, numLinks());
  Jst_eef.col(0) = screw_axes[0];
  for (int i = 1; i < numLinks(); ++i) {
    Jst_eef.col(i) = transforms[i - 1].AdjointMap() * screw_axes[i];
  }

  for (int i = 0; i < numLinks(); ++i) {
    insertSub(Js[i], sub(Jst_eef, 0, 6, 0, i + 1), 0, 0);
  }
  Js[numLinks()] = Jst_eef;

  return Js;
}

template <typename T>
vector<Matrix> SerialLink<T>::bodyManipulatorJacobian(
    const Vector &q, vector<Pose3> &sTb) const {
  // Calculate spatial_manipulator_jacobian
  vector<Matrix> Js = spatialManipulatorJacobian(q);
  // assign space for manipulator jacobian
  vector<Matrix> Jb(numLinks() + 1, Matrix::Zero(6, numLinks()));
  // Convert spatial_manipulator_jacobian to body_manipulator_jacobian,
  // Lynch & Park's book page 187, equation (5.22)
  for (int i = 0; i <= numLinks(); ++i) {
    Jb[i] = sTb[i].inverse().AdjointMap() * Js[i];
  }
  return Jb;
}

template <typename T>
vector<Vector6> SerialLink<T>::twists(const std::vector<Pose3> &Ts,
                                      const Vector &joint_velocities) const {
  vector<Vector6> twists;
  twists.push_back(screwAxes_[0] * joint_velocities[0]);
  Vector6 twist_i;
  Vector6 twist_j;
  Vector6 Aj;
  double joint_vel_j;
  Pose3 jTi;

  for (int j = 2; j <= numLinks(); ++j) {
    twist_i = twists.back();
    jTi = Ts[j - 1].between(Ts[j - 2]);
    Aj = screwAxes_[j - 1];
    joint_vel_j = joint_velocities[j - 1];
    twist_j = jTi.AdjointMap() * twist_i + Aj * joint_vel_j;
    twists.push_back(twist_j);
  }
  return twists;
}

template <typename T>
vector<Pose3> SerialLink<T>::jTi_list(const Vector &q) const {
  vector<Pose3> Ts = comFrames(q);
  Pose3 bT1 = Ts[0].between(base_);
  Pose3 nTt = tool_;
  Pose3 nTc = links_.back().centerOfMass();
  Pose3 tTnc = nTt.between(nTc);
  vector<Pose3> jTi_com;
  jTi_com.push_back(bT1);
  for (int j = 1; j < numLinks(); ++j) {
    jTi_com.push_back(Ts[j].between(Ts[j - 1]));
  }
  jTi_com.push_back(tTnc);
  return jTi_com;
}

template <typename T>
GaussianFactorGraph SerialLink<T>::forwardDynamicsFactorGraph(
    const Vector &q, const Vector &joint_velocities, const Vector &torques,
    const Vector6 &base_twist_accel, const Vector6 &external_wrench,
    boost::optional<Vector3 &> gravity) const {
  int N = numLinks();
  assert(q.size() == N);
  assert(joint_velocities.size() == N);
  assert(torques.size() == N);

  // Configuration of COM link frames
  vector<Pose3> Ts = comFrames(q);
  // Calculate all twists
  vector<Vector6> twists_vec = twists(Ts, joint_velocities);
  // Set up Gaussian Factors Graph
  GaussianFactorGraph gfg = GaussianFactorGraph();
  // Add factor to enforce base acceleration
  gfg.add(Link::BaseTwistAccelFactor(base_twist_accel));
  // Configuration of link frame j-1 relative to link frame j for arbitrary
  // joint angle
  vector<Pose3> jTis = jTi_list(q);

  int j = 0;
  for (int i = 0; i < N; ++i) {
    j = i + 1;
    auto jRw = Ts[i].rotation().inverse();
    if (gravity) {
      gfg.push_back(links_[i].forwardFactors(j, jTis[i], joint_velocities[i],
                                             twists_vec[i], torques[i], jTis[j],
                                             jRw * (*gravity)));
    } else {
      gfg.push_back(links_[i].forwardFactors(j, jTis[i], joint_velocities[i],
                                             twists_vec[i], torques[i], jTis[j],
                                             gravity));
    }
  }
  // Add factor to enforce external wrench at tool
  gfg.add(Link::ToolWrenchFactor(N, external_wrench));

  return gfg;
}

template <typename T>
GaussianFactorGraph SerialLink<T>::inverseDynamicsFactorGraph(
    const Vector &q, const Vector &joint_velocities,
    const Vector &joint_accelerations, const Vector6 &base_twist_accel,
    const Vector6 &external_wrench, boost::optional<Vector3 &> gravity) const {
  int N = numLinks();
  assert(q.size() == N);
  assert(joint_velocities.size() == N);
  assert(joint_accelerations.size() == N);

  // Configuration of COM link frames
  vector<Pose3> Ts = comFrames(q);
  // Calculate all twists
  vector<Vector6> twists_vec = twists(Ts, joint_velocities);
  // Set up Gaussian Factors Graph
  GaussianFactorGraph gfg = GaussianFactorGraph();
  // Add factor to enforce base acceleration
  gfg.add(Link::BaseTwistAccelFactor(base_twist_accel));
  // Configuration of link frame j-1 relative to link frame j for arbitrary
  // joint angle
  vector<Pose3> jTis = jTi_list(q);

  int j = 0;
  for (int i = 0; i < N; ++i) {
    j = i + 1;
    auto jRw = Ts[i].rotation().inverse();
    if (gravity) {
      gfg.push_back(links_[i].inverseFactors(
          j, jTis[i], joint_velocities[i], twists_vec[i],
          joint_accelerations[i], jTis[j], jRw * (*gravity)));
    } else {
      gfg.push_back(links_[i].inverseFactors(
          j, jTis[i], joint_velocities[i], twists_vec[i],
          joint_accelerations[i], jTis[j], gravity));
    }
  }
  // Add factor to enforce external wrench at tool
  gfg.add(Link::ToolWrenchFactor(N, external_wrench));

  return gfg;
}

template <typename T>
Vector SerialLink<T>::extractJointAcceleraions(
    const VectorValues &result) const {
  int N = numLinks();
  Vector joint_accels(N);
  for (int j = 1; j <= N; ++j) {
    joint_accels[j - 1] = result.at(a(j))(0);
  }
  return joint_accels;
}

template <typename T>
Vector SerialLink<T>::extractTorques(const VectorValues &result) const {
  int N = numLinks();
  Vector torques(N);
  for (int j = 1; j <= N; ++j) {
    torques[j - 1] = result.at(t(j))(0);
  }
  return torques;
}

template <typename T>
VectorValues SerialLink<T>::factorGraphOptimization(
    const GaussianFactorGraph &dynamics_factor_graph) const {
  return dynamics_factor_graph.optimize();
}

template <typename T>
Vector SerialLink<T>::forwardDynamics(
    const Vector &q, const Vector &joint_velocities, const Vector &torques,
    const Vector6 &base_twist_accel, const Vector6 &external_wrench,
    boost::optional<Vector3 &> gravity) const {
  GaussianFactorGraph factor_graph = forwardDynamicsFactorGraph(
      q, joint_velocities, torques, base_twist_accel, external_wrench, gravity);
  VectorValues result = factor_graph.optimize();
  return extractJointAcceleraions(result);
}

template <typename T>
Vector SerialLink<T>::inverseDynamics(
    const Vector &q, const Vector &joint_velocities,
    const Vector &joint_accelerations, const Vector6 &base_twist_accel,
    const Vector6 &external_wrench, boost::optional<Vector3 &> gravity) const {
  GaussianFactorGraph factor_graph =
      inverseDynamicsFactorGraph(q, joint_velocities, joint_accelerations,
                                 base_twist_accel, external_wrench, gravity);
  VectorValues result = factor_graph.optimize();
  return extractTorques(result);
}

template <typename T>
JointLimitVectorFactor SerialLink<T>::jointLimitVectorFactor() const {
  noiseModel::Gaussian::shared_ptr cost_model =
      noiseModel::Isotropic::Sigma(numLinks(), 1.0);
  return JointLimitVectorFactor(J(0), cost_model, jointLowerLimits(),
                                jointUpperLimits(), jointLimitThresholds());
}

template <typename T>
PoseGoalFactor SerialLink<T>::poseGoalFactor(const Pose3 &goal_pose) const {
  noiseModel::Gaussian::shared_ptr cost_model =
      noiseModel::Isotropic::Sigma(6, 1.0);
  return PoseGoalFactor(
      J(0), cost_model, goal_pose,
      boost::bind(&SerialLink<T>::forwardKinematics, *this, _1, _2));
}

template <typename T>
NonlinearFactorGraph SerialLink<T>::inverseKinematicsFactorGraph(
    const Pose3 &goal_pose) const {
  NonlinearFactorGraph graph;
  graph.add(poseGoalFactor(goal_pose));
  graph.add(jointLimitVectorFactor());
  return graph;
}

template <typename T>
Vector SerialLink<T>::extractJointCooridinates(const Values &results) const {
  return results.at<Vector>(J(0));
}

template <typename T>
Values SerialLink<T>::factorGraphOptimization(const NonlinearFactorGraph &graph,
                                              const Values &init_values) const {
  LevenbergMarquardtOptimizer optimizer(graph, init_values);
  optimizer.optimize();
  return optimizer.values();
}

template <typename T>
Vector SerialLink<T>::inverseKinematics(const Pose3 &goal_pose,
                                        const Vector &init_q) const {
  NonlinearFactorGraph graph = inverseKinematicsFactorGraph(goal_pose);
  Values init_values;
  init_values.insert(J(0), init_q);
  Values results = factorGraphOptimization(graph, init_values);
  return extractJointCooridinates(results);
}

template class SerialLink<DH_Link>;
template class SerialLink<URDF_Link>;
}  // namespace manipulator
